### PR TITLE
linkerd module

### LINKED ISSUE #257: Add linkerd module
Todo:

* [x] Add linkerd module
* [x] Add linkerd to `aks-core` as optional
* [x] Update `ingress-nginx` to work with linkerd
* [x] Update `ingress-healthz` to inject linkerd-proxy

Requirements:

* [x] Linkerd must be optional
* [x] Should be enabled per deployment, opt-in
* [x] ~Ingress must work with and without linkerd sidecar or an additional ingress is required for linkerd enabled services~

Additional information:

- `ingress-nginx` needs linkerd-proxy to get mTLS to services with linkerd-proxy, but still works with services that doesn't have it
- `ingress-nginx` required `config.linkerd.io/skip-inbound-ports: "80,443"` to work with `nginx.ingress.kubernetes.io/whitelist-source-range` (more info [here](https://github.com/linkerd/linkerd2/issues/3334#issuecomment-565135188))
- added proxy injection to `ingress-healthz` to make sure it includes validation of mTLS
- it is a manual process including a deployment / pod in the mesh, using the annotation `linkerd.io/inject: enabled` (more info [here](https://linkerd.io/2.10/tasks/adding-your-service/))
- to get ingress working for a service, additional configuration is needed (more info [here](https://linkerd.io/2.10/tasks/using-ingress/#nginx))
- to get linkerd to work with the `opa-gatekeeper` rules applied, [linkerd-cni](https://linkerd.io/2.10/features/cni/) was required

### PR DESCRIPTION
**Pull request by simongottschlag**
Info:
- `ingress-nginx` needs linkerd-proxy to get mTLS to services with linkerd-proxy, but still works with services that doesn't have it
- `ingress-nginx` required `config.linkerd.io/skip-inbound-ports: "80,443"` to work with `nginx.ingress.kubernetes.io/whitelist-source-range` (more info [here](https://github.com/linkerd/linkerd2/issues/3334#issuecomment-565135188))
- added proxy injection to `ingress-healthz` to make sure it includes validation of mTLS
- it is a manual process including a deployment / pod in the mesh, using the annotation `linkerd.io/inject: enabled` (more info [here](https://linkerd.io/2.10/tasks/adding-your-service/))
- to get ingress working for a service, additional configuration is needed (more info [here](https://linkerd.io/2.10/tasks/using-ingress/#nginx))
- to get linkerd to work with the `opa-gatekeeper` rules applied, [linkerd-cni](https://linkerd.io/2.10/features/cni/) was required

Closes #257

### CODE REVIEW

**Comment by simongottschlag** [comment]
@phillebaba @NissesSenap ready for review :+1:

File: modules/kubernetes/ingress-healthz/templates/values.yaml.tpl [hunk 1]
```diff
@@ -28,11 +28,21 @@ ingress:
   hostname: ingress-healthz.${dns_zone}
   annotations:
     cert-manager.io/cluster-issuer: letsencrypt
+    %{ if linkerd_enabled }
+    nginx.ingress.kubernetes.io/configuration-snippet: |
+      proxy_set_header l5d-dst-override $service_name.$namespace.svc.cluster.local:$service_port;
+      grpc_set_header l5d-dst-override $service_name.$namespace.svc.cluster.local:$service_port;
```
Comment (phillebaba) [inline modules/kubernetes/ingress-healthz/templates/values.yaml.tpl:34]: So do all ingress resources need this when linkerd is enabled? Even when they are not using the sidecar? Could be solve with with a mutation?

File: modules/kubernetes/linkerd/main.tf [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    "tls.crt" = tls_self_signed_cert.webhook_issuer_tls.cert_pem
+    "tls.key" = tls_private_key.webhook_issuer_tls.private_key_pem
+  }
+
+  type = "kubernetes.io/tls"
+}
+
+# Install linkerd-cni helm chart
+locals {
+  values_cni = templatefile("${path.module}/templates/values-cni.yaml.tpl", {})
```
Comment (phillebaba) [inline modules/kubernetes/linkerd/main.tf:166]: Can we inline the values instead of using locals.

File: modules/kubernetes/linkerd/main.tf [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    kubernetes_secret.webhook_issuer_tls
+  ]
+
+  chart     = "${path.module}/charts/linkerd-extras"
+  name      = "linkerd-extras"
+  namespace = kubernetes_namespace.this.metadata[0].name
+}
+
+locals {
+  values = templatefile("${path.module}/templates/values.yaml.tpl", {
```
Comment (phillebaba) [inline modules/kubernetes/linkerd/main.tf:190]: ditto

File: modules/kubernetes/ingress-healthz/templates/values.yaml.tpl (hunk 1 above)
Comment (simongottschlag) [reply modules/kubernetes/ingress-healthz/templates/values.yaml.tpl:34]: These two headers are always needed if proxy injection is enabled. My opinion as of right now is that I don't want to hide away this complexity from the developer but instead make them understand it. If it becomes a problem, perhaps we can create a mutation for it - but another PR.

Do you agree?

File: modules/kubernetes/linkerd/main.tf (hunk 2 above)
Comment (simongottschlag) [reply modules/kubernetes/linkerd/main.tf]: Done!

File: modules/kubernetes/linkerd/main.tf (hunk 3 above)
Comment (simongottschlag) [reply modules/kubernetes/linkerd/main.tf]: Done!

File: modules/kubernetes/ingress-healthz/templates/values.yaml.tpl (hunk 1 above)
Comment (simongottschlag) [reply modules/kubernetes/ingress-healthz/templates/values.yaml.tpl:34]: I've also updated the README with more info regarding this.

File: modules/kubernetes/linkerd/charts/linkerd-extras/templates/wait-for-certs.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
+apiVersion: v1
```
Comment (phillebaba) [inline modules/kubernetes/linkerd/charts/linkerd-extras/templates/wait-for-certs.yaml:1]: Is this still needed when the secret is created with Terraform?

File: modules/kubernetes/ingress-healthz/templates/values.yaml.tpl (hunk 1 above)
Comment (phillebaba) [reply modules/kubernetes/ingress-healthz/templates/values.yaml.tpl:34]: Sounds good.

File: modules/kubernetes/linkerd/charts/linkerd-extras/templates/wait-for-certs.yaml (hunk 4 above)
Comment (simongottschlag) [reply modules/kubernetes/linkerd/charts/linkerd-extras/templates/wait-for-certs.yaml:1]: I want to make sure all certificates are issued before we install the helm chart for linkerd, making sure that we don't start the control plane before they are issued.

Sounds OK?

Comment (simongottschlag) [reply modules/kubernetes/linkerd/charts/linkerd-extras/templates/wait-for-certs.yaml]: Removed it according to call.

**Review by phillebaba — APPROVED** [review]

**Pull request merged**